curve_test.go

```diff
@@ -146,3 +152,248 @@ func TestMultAir(t *testing.T) {
 		}
 	}
 }
+
+// swappable ec multiplication fn
+type ecMultiFn func(m, x1, y1 *big.Int) (x, y *big.Int)
+type ecMultOption struct {
+	algo   string
+	fn     ecMultiFn
+	stddev float64
+}
+
+// Get multiple ec multiplication algo options to test and benchmark
+func (sc StarkCurve) ecMultOptions() []ecMultOption {
+	return []ecMultOption{
+		{
+			algo: "Add-Always-Double",
+			fn:   sc.EcMult, // used algo
+		},
+		{
+			algo: "Double-And-Add",
+			fn:   sc.ecMult_DoubleAndAdd, // original algo
+		},
+		{
+			algo: "Montgomery-Ladder",
+			fn:   sc.ecMult_Montgomery,
+		},
+		{
+			algo: "Montgomery-Ladder-Lsh",
+			fn:   sc.ecMult_MontgomeryLsh,
+		},
+	}
+}
+
+// obtain public key coordinates from stark curve given the private key (configurable ec multlipication fn)
+func (sc StarkCurve) privateToPoint(privKey *big.Int, ecMulti ecMultiFn) (x, y *big.Int, err error) {
+	if privKey.Cmp(big.NewInt(0)) != 1 || privKey.Cmp(sc.N) != -1 {
+		return x, y, fmt.Errorf("private key not in curve range")
+	}
+	x, y = ecMulti(privKey, sc.EcGenX, sc.EcGenY)
+	return x, y, nil
+}
+
+func TestEcMult(t *testing.T) {
+	tests := []struct {
+		k         *big.Int
+		expectedX *big.Int
+		expectedY *big.Int
+	}{
+		{
+			k:         StrToBig("1"),
+			expectedX: StrToBig("874739451078007766457464989774322083649278607533249481151382481072868806602"),
+			expectedY: StrToBig("152666792071518830868575557812948353041420400780739481342941381225525861407"),
+		},
+		{
+			k:         StrToBig("2"),
+			expectedX: StrToBig("3324833730090626974525872402899302150520188025637965566623476530814354734325"),
+			expectedY: StrToBig("3147007486456030910661996439995670279305852583596209647900952752170983517249"),
+		},
+		{
+			k:         StrToBig("3"),
+			expectedX: StrToBig("1839793652349538280924927302501143912227271479439798783640887258675143576352"),
+			expectedY: StrToBig("3564972295958783757568195431080951091358810058262272733141798511604612925062"),
+		},
+		{
+			k:         StrToBig("2458502865976494910213617956670505342647705497324144349552978333078363662855"),
+			expectedX: StrToBig("3547379334427783199328848231018312737594618983118903125858401905055740589204"),
+			expectedY: StrToBig("1130045581446120977306267646220500675898605907506384721505308463662654640404"),
+		},
+		{
+			k:         StrToBig("2647705497324144349552978333078363662855245850286597649491021361795667050534"),
+			expectedX: StrToBig("1976290247577832044445658844239552133924048365220576311475124074782390231872"),
+			expectedY: StrToBig("375316965946589635161928898648099004919185843055494241447142141522927320685"),
+		},
```

Comment:
Where did you get these test vectors? I'd add some tests that do `curve.Sign`/`curve.Verify` against known good signature values/keypairs. This is because we've seen issues with `juno` that was able to sign and verify it's own signatures, but failed to verify signatures generated by the network.

Reply:
I generated them. They were tested against the original implementation, and then added as test vectors to verify all implementations behave the same. 

Now, the test suite does have a few signature verifications it tests, and those currently pass. Do you know where to find extra good signature/key pairs? Maybe Juno has some...

What I want to do next is to add a fuzzer for this multiple EC multiplication fn test.

Reply:
`juno` had issues since it was testing against itself, maybe we can find some tests from the `cairo` repository

Reply:
I've added the example test from here: https://github.com/starkware-libs/crypto-cpp/blob/master/src/starkware/crypto/ecdsa_test.cc

Didn't find any other valid examples (Juno has none). If you know where to source some, let me know.